src/Course/StateT.hs

```diff
@@ -194,27 +152,24 @@ distinct' =
 --
 -- >>> distinctF $ listh [1,2,3,2,1,101]
 -- Empty
-distinctF ::
-  (Ord a, Num a) =>
-  List a
-  -> Optional (List a)
-distinctF =
-  error "todo: Course.StateT#distinctF"
+distinctF :: (Ord a, Num a) => List a -> Optional (List a)
+distinctF xs = reverse <$> evalT (filtering p xs) S.empty
+  where
+    p x
+      | x < 100 = StateT (\s -> Full (S.notMember x s, S.insert x s))
+      | otherwise = StateT (const Empty)
+
+-- distinctF = filtering (\x -> pure (x < 100)) . distinct'
 
 -- | An `OptionalT` is a functor of an `Optional` value.
-data OptionalT f a =
-  OptionalT {
-    runOptionalT ::
-      f (Optional a)
-  }
+data OptionalT f a = OptionalT {runOptionalT :: f (Optional a)}
 
 -- | Implement the `Functor` instance for `OptionalT f` given a Functor f.
 --
 -- >>> runOptionalT $ (+1) <$> OptionalT (Full 1 :. Empty :. Nil)
 -- [Full 2,Empty]
 instance Functor f => Functor (OptionalT f) where
-  (<$>) =
-    error "todo: Course.StateT (<$>)#instance (OptionalT f)"
+  f <$> (OptionalT funct) = OptionalT (((<$>) . (<$>)) f funct)
```

Comment:
👍 I quite liked greg's implementation he did for this. I did a similar thing to you.

Reply:
definitely (fmap . fmap) looks nicer, but we don't have that one